wiktextract/extractor/zh/page.py

```diff
@@ -0,0 +1,293 @@
+            ],
+        )
+        extract_sense_tags(wxr, page_data, gloss)
+
+
+def extract_sense_tags(
+    wxr: WiktextractContext, page_data: List[Dict], raw_gloss: str
+) -> None:
+    if raw_gloss.startswith("("):
+        label_end = raw_gloss.find(")")
+        label = raw_gloss[1:label_end]
+        gloss = raw_gloss[label_end + 2 :]  # also remove space after ")"
+        # labels: https://zh.wiktionary.org/wiki/Module:Labels/data
+        tags = re.split(r",|或", label)
+        page_data[-1]["senses"].append(
+            {"glosses": [gloss], "raw_glosses": [raw_gloss], "tags": tags}
+        )
+    else:
+        page_data[-1]["senses"].append({"glosses": [raw_gloss]})
+
+
+def extract_etymology(
+    wxr: WiktextractContext,
+    page_data: List[Dict],
+    base_data: Dict,
+    nodes: List[Union[WikiNode, str]],
+) -> None:
+    for index, node in enumerate(nodes):
+        if isinstance(node, WikiNode) and node.kind == NodeKind.TEMPLATE:
+            etymology = clean_node(
+                wxr,
+                None,
+                [
+                    child
+                    for child in node.children
+                    if not isinstance(child, WikiNode)
+                    or (
+                        child.kind != NodeKind.LIST
+                        and child.kind not in LEVEL_KINDS
+                    )
+                ],
+            )
+            base_data["etymology_text"] = etymology
+            append_dict(page_data, "etymology_text", etymology, base_data)
+            recursive_parse(wxr, page_data, base_data, nodes[index + 1 :])
+            return
+
+
+def extract_pronunciation(
+    wxr: WiktextractContext,
+    page_data: List[Dict],
+    base_data: Dict,
+    nodes: List[Union[WikiNode, str]],
+) -> None:
+    for index, node in enumerate(nodes):
+        if isinstance(node, WikiNode) and node.kind == NodeKind.TEMPLATE:
+            # extract ipa
+            base_data["sounds"] = "sounds_placeholder"
+            append_dict(page_data, "sounds", "sounds_placeholder", base_data)
+            recursive_parse(wxr, page_data, base_data, nodes[index + 1 :])
+            return
+
+
+def parse_page(
+    wxr: WiktextractContext, page_title: str, page_text: str
+) -> List[Dict[str, str]]:
+    if wxr.config.verbose:
+        logging.info(f"Parsing page: {page_title}")
+
+    wxr.config.word = page_title
+    wxr.wtp.start_page(page_title)
+
+    # Parse the page, pre-expanding those templates that are likely to
+    # influence parsing
+    tree = wxr.wtp.parse(
+        page_text,
+        pre_expand=True,
+        additional_expand=ADDITIONAL_EXPAND_TEMPLATES,
+    )
+
+    page_data = []
+    for node in filter(
+        # ignore link created by `also` template at the page top
+        lambda n: isinstance(n, WikiNode) and n.kind != NodeKind.LINK,
+        tree.children,
+    ):
+        if node.kind != NodeKind.LEVEL2:
+            logging.warning(f"Unexpected top-level node: {node}")
```

Comment:
Should migrate to Wtp.debug(), .warning() and .error() for compatibility with our other logging functions.

Reply:
Sure, I'll change them. I guess the `sortid` argument of `Wtp.error()` is used to group and sort errors on kaikki.org? So it's OK to use outdated line number as long as it's unique?

Reply:
That is correct, it's just a handy way to sort them. Originally, I used a costly function to check the traceback and find where the logging call was *from*, but that was too expensive so I just added manual string parameters instead.